packages/accounts-password/password_server.js

```diff
@@ -239,8 +239,22 @@ const checkForCaseInsensitiveDuplicates = (fieldName, displayName, fieldValue, o
   const skipCheck = Object.prototype.hasOwnProperty.call(Accounts._skipCaseInsensitiveChecksForTest, fieldValue);
 
   if (fieldValue && !skipCheck) {
+    let selector = {};
+    const { useLowerCaseEmails } = Accounts._options;
+    // if useLowerCaseEmails is false in Accounts.config and fieldName='emails.address'
+    // then run selectorForFastCaseInsensitiveLookup method
+    // else run a simple regex query for email.
+    if (fieldName === 'emails.address') {
+      if(useLowerCaseEmails === false) {
+        selector = selectorForFastCaseInsensitiveLookup(fieldName, fieldValue);
+      } else {
+        selector[fieldName] = {$regex: fieldValue, $options: 'i'};
+      }
```

Comment:
If we're sure the db emails are lower case(*), and the query email can be converted to lowercase, why not just use a simple match `selector[fieldName] = fieldValue.toLowerCase()`?  Isn't that the whole point of this PR?

(*) we can only be sure that all exiting db emails are lower case if we make this to be an opt-in behaviour, and inform the developer that they must migrate the db before opting in...

Reply:
> Have you compared the performance for both of these queries? In the past, `selectorForFastCaseInsensitiveLookup` was much faster than using a case insentive regexp by itself since it isn't able to use indexes. Both queries should produce the same results, so it should be fine to use the the faster query in both cases.

Hi  @zodern,
I did the bench-marking tests for this query.
I created two meteor apps with the with this pr's build, one using **useLowerCaseEmails = true** and other using **useLowerCaseEmails = false** in Accounts.config.
I used [performance.now() ](https://developer.mozilla.org/en-US/docs/Web/API/Performance/now)to log the time spent on creating new a user in both the meteor apps. I do not see significant time difference after each call to Accounts.createUser.
Below is the code snippet  that I ran on client side in the app. This code was basically creating random users in both the application at an interval of 7 seconds.
 ```
setInterval( async ()=>{
    const id = Random.id();
    const email = ritwik-tesT${id}@gmail.com
    const t0 = performance.now();
    await Accounts.createUser({
      email,
      password: 'test-password'
    });
    const t1 = performance.now();
    console.log(Call took ${t1 - t0} milliseconds.);
  }, 7000);
```
But one thing I noticed is that after running both apps locally for around 30 minutes continuously, the app with **useLowerCaseEmails = false** created around **529** users while the app with **useLowerCaseEmails = true** created around **674** users.

Not sure If this is a good bench marking test, but I think this difference is caused by **selectorForFastCaseInsensitiveLookup** method being used in the app with **useLowerCaseEmails = false** and the emails starting with common first name.

Reply:
I created a couple scripts that can be run in the mongo shell to help test performance: https://gist.github.com/zodern/4ad546b9ddc2cf780fed476c85c1f053

The first one `generate-users.js` generates the emails and inserts them into the `usersTest` collection. `GROUPS` is the number of different combinations for the first 4 characters used in the generated emails.

`run-queries.js` samples 10 emails from the collection, adds a suffix to each one to make it unique, and then prints how long queries took for the emails.

I didn't run too many tests, but it seems as the number of emails or groups increases `selectorForFastCaseInsensitiveLookup ` becomes more efficient since it is able to use indexes to exclude emails that don't have the same first 4 characters, whereas the case insensitive regexp always does a collection scan. When there is only one group, the case insensitive regexp seems to be faster.

Reply:
> If we're sure the db emails are lower case(*), and the query email can be converted to lowercase, why not just use a simple match selector[fieldName] = fieldValue.toLowerCase()? Isn't that the whole point of this PR?

This would work and fix the performance, though it is a little more complicated since apps with users created before Meteor 1.2 might have multiple users with emails that only differ by case, and when they login Meteor still treats the email as case sensitive.